File: c++/remove_duplicate_letters.cpp
/*
Given a string which contains only lowercase letters,
remove duplicate letters so that every letter appear
once and only once. You must make sure your result
is the smallest in lexicographical order among all
possible results.

Example:
Given "bcabc"
Return "abc"

Given "cbacdcbc"
Return "acdb"
*/

#include <string>
#include <iostream>

using namespace std;

class Solution {
public:
    /**
     * Count the frequency of letters in the string.
     * Loop through the string, each time,
     * 1. if the current string is already in the result string, continue
     * 2. if the current string is not used, compare it with the back
     * of the result string. If smaller and the back char has remaining
     * frequency, it means we can pop back char to make the result string
     * lexically smaller. Keep doing this until either the result string
     * is empty or the current char is greater than the back char.
     */
    string removeDuplicateLetters(string s) {
        int frequency[26] = {0};
        bool visited[26] = {false};
        string str;
        for (auto c : s) ++frequency[(int)(c-'a')];
        for (auto c : s) {
            int idx = (int)(c - 'a');
            --frequency[idx];
            if (visited[idx]) continue;
            while (!str.empty() && c < str.back() && frequency[(int)(str.back()-'a')] > 0) {
                visited[(int)(str.back()-'a')] = false;
                str.pop_back();
            }
            str.push_back(c);
            visited[idx] = true;
        }
        return str;
    }
};

File: java/RemoveDuplicateLetters.java
/**
 * Created on 8/20/16.
 * Given a string which contains only lowercase letters,
 * remove duplicate letters so that every letter appear
 * once and only once. You must make sure your result
 * is the smallest in lexicographical order among all
 * possible results.
 *
 * Example:
 * Given "bcabc"
 * Return "abc"
 *
 * Given "cbacdcbc"
 * Return "acdb"
 */

public class RemoveDuplicateLetters {

    public static void main(String[] args) {
        String s = "cbacdcbc";
        RemoveDuplicateLetters app = new RemoveDuplicateLetters();
        System.out.println(app.removeDuplicateLetters(s));
    }

    public String removeDuplicateLetters(String s) {
        int[] frequency = new int[26];
        for (int i = 0; i < s.length(); ++i) {
            ++frequency[s.charAt(i)-'a'];
        }
        boolean[] used = new boolean[26];
        StringBuilder str = new StringBuilder();
        for (int i = 0; i < s.length(); ++i) {
            char c = s.charAt(i);
            int idx = c - 'a';
            --frequency[idx];
            if (used[idx]) continue;
            while (str.length() != 0 && c < str.charAt(str.length()-1) && frequency[str.charAt(str.length()-1)-'a'] > 0) {
                used[str.charAt(str.length()-1)-'a'] = false;
                str.deleteCharAt(str.length()-1);
            }
            str.append(c);
            used[idx] = true;
        }
        return str.toString();
    }
}
